libsolidity/experimental/codegen/IRGeneratorForStatements.cpp

```diff
@@ -201,13 +201,50 @@ void IRGeneratorForStatements::endVisit(BinaryOperation const& _binaryOperation)
 	Type functionType = helper.functionType(helper.tupleType({leftType, rightType}), resultType);
 	auto [typeClass, memberName] = m_context.analysis.annotation<TypeInference>().operators.at(_binaryOperation.getOperator());
 	auto const& functionDefinition = resolveTypeClassFunction(typeClass, memberName, functionType);
-	// TODO: deduplicate with FunctionCall
+	std::string functionDeclaration = var(_binaryOperation).commaSeparatedList();
+	if (!functionDeclaration.empty())
+		m_code << "let " << functionDeclaration << " := ";
+	m_code << buildFunctionCall(functionDefinition, functionType, _binaryOperation.arguments());
+}
+
+std::string IRGeneratorForStatements::buildFunctionCall(FunctionDefinition const& _functionDefinition, Type _functionType, std::vector<ASTPointer<Expression const>> const& _arguments)
+{
+	// Ensure type is resolved
 	// TODO: get around resolveRecursive by passing the environment further down?
-	functionType = m_context.env->resolveRecursive(functionType);
-	m_context.enqueueFunctionDefinition(&functionDefinition, functionType);
-	// TODO: account for return stack size
-	m_code << "let " << IRNames::localVariable(_binaryOperation) << " := " << IRNames::function(*m_context.env, functionDefinition, functionType) << "("
-		<< IRNames::localVariable(_binaryOperation.leftExpression()) << ", " << IRNames::localVariable(_binaryOperation.rightExpression()) << ")\n";
+	Type resolvedFunctionType = m_context.env->resolveRecursive(_functionType);
+	m_context.enqueueFunctionDefinition(&_functionDefinition, resolvedFunctionType);
+
+	std::ostringstream output;
+	output << IRNames::function(*m_context.env, _functionDefinition, resolvedFunctionType) << "(";
+	if (_arguments.size() == 1)
+		output << var(*_arguments.back()).commaSeparatedList();
+	else if (_arguments.size() > 1)
+	{
+		for (auto arg: _arguments | ranges::views::drop_last(1))
+			output << var(*arg).commaSeparatedList();
+		output << var(*_arguments.back()).commaSeparatedListPrefixed();
+	}
+	output << ")\n";
+	return output.str();
+}
+
+void IRGeneratorForStatements::assign(IRVariable const& _lhs, IRVariable const& _rhs, bool _declare)
+{
+	solAssert(IRGeneratorForStatements::stackSize(m_context, _lhs.type()) == IRGeneratorForStatements::stackSize(m_context, _rhs.type()));
+	// TODO: conversions for abs and rep for user-defined types as no-ops
+	if (IRGeneratorForStatements::stackSize(m_context, _lhs.type()) == 1)
+		m_code << (_declare ? "let " : "") <<  _lhs.name() << " := " << _rhs.name() << "\n";
+	else
+	{
+		for (size_t i = 0; i < _lhs.stackSize(); ++i)
+			m_code << (_declare ? "let " : "") << _lhs.stackSlots()[i] << " := " << _rhs.stackSlots()[i] << "\n";
```

Comment:
```suggestion
		auto lhsSlots = _lhs.stackSlots();
		auto rhsSlots = _rhs.stackSlots();
		for (size_t i = 0; i < _lhs.stackSize(); ++i)
			m_code << (_declare ? "let " : "") << lhsSlots[i] << " := " << rhsSlots[i] << "\n";
```
otherwise you're creating a new vector of slots in each loop iteration...

Actually we might also be able to get fancy and do something like
```suggestion
		for (auto&& [lhsSlot, rhsSlot]: ranges::zip_view(_lhs.stackSlots(), _rhs.stackSlots()))
			m_code << (_declare ? "let " : "") << lhsSlot << " := " <<  rhsSlot << "\n";
```
or something similar to that

Reply:
Nice ;)